core/ibft.go

```diff
@@ -558,28 +552,25 @@ func (i *IBFT) runNewRound(ctx context.Context) error {
 	defer i.messages.Unsubscribe(sub.ID)
 
 	for {
-		select {
-		case <-ctx.Done():
-			// Stop signal received, exit
-			return errTimeoutExpired
-		case <-sub.SubCh:
-			// SubscriptionDetails conditions have been met,
-			// grab the proposal messages
-			proposalMessage := i.handlePrePrepare(view)
-			if proposalMessage == nil {
-				continue
-			}
-
+		// SubscriptionDetails conditions have been met,
+		// grab the proposal messages
+		proposalMessage := i.handlePrePrepare(view)
+		if proposalMessage != nil {
 			// Multicast the PREPARE message
-			i.state.setProposalMessage(proposalMessage)
+			i.acceptProposal(proposalMessage)
 			i.sendPrepareMessage(view)
 
 			i.log.Debug("prepare message multicasted")
 
-			// Move to the prepare state
-			i.state.changeState(prepare)
+			return
+		}
 
-			return nil
+		select {
+		case <-ctx.Done():
```

Comment:
why we'd like to exit in the end, not in the begining?

Reply:
https://github.com/0xPolygon/go-ibft/pull/57#discussion_r1087796922